### DISCUSSION TITLE
How to compile with external jars?

### DISCUSSION THREAD

**Discussion by programmer1o1** [MCPHackers/RetroMCP-Java#100] [Q&A]
I know there is a wiki on adding libraries but that was edited in 2022 and it probably changed. So, I wanna know how to compile with external jars. I cannot seem to find the `lib` folder. Sorry, if you find this question stupid.

**Comment by AZ-1282** [comment]
i am trying to get forge 1.5.2 to decompile but the error gives me a missing library even when i put the forge libraries next to the game's ones, idk where to put those either

**Reply by PhoenixVX** [reply]
Forge is expected to be decompiled as a JAR mod of the standard `minecraft.jar`. Take the classes from the Forge ZIP and use 7zip/WinRAR to apply those classes onto the vanilla JAR. Remove the META-INF folder inside `minecraft.jar`.

Additionally, Forge decompiles better when you turn on the "Strip generics" option which also strips out the local variable table which may have name collisions with the auto generated Fernflower names.

**Comment by PhoenixVX** [comment]
The approach here depends on whether or not the library is needed to be reobfuscated or bundled with your game JAR. 

If the library needs to be bundled with your JAR and reobfuscated, you can follow the advice I gave above with Forge.

If the library does not need to be bundled or reobfuscated, use the "Java compiler arguments" option to add new JARs to the compile classpath via `-cp "path/to/library.jar"`
